lib/mix/lib/mix/tasks/compile.ex

```diff
@@ -90,11 +90,19 @@ defmodule Mix.Tasks.Compile do
 
     load_erl_config(opts)
 
-    {res, diagnostics} =
+    diagnostics =
       Mix.Task.run("compile.all", args)
       |> List.wrap()
       |> Enum.map(&Mix.Task.Compiler.normalize(&1, :all))
-      |> Enum.reduce(&merge_diagnostics/2)
+
+    {res, diagnostics} =
+      case diagnostics do
+        [] ->
+          Enum.reduce(diagnostics, {:ok, []}, &merge_diagnostics/2)
```

Comment:
We can keep the code exactly the same as before, we just need to change one line where we replace Enum.reduce/2 by a Enum.reduce/3. :)

Reply:
I need to change old tests. Do you agree?
Assertion with == failed
     code:  assert Mix.Task.run("compile", ["--verbose"]) == {:noop, []}
     left:  {:ok, []}
     right: {:noop, []}
     stacktrace:
       lib/mix/test/mix/tasks/compile_test.exs:54: anonymous fn/0 in Mix.Tasks.CompileTest."test compile a project with mixfile"/1
       (elixir) lib/file.ex:1428: File.cd!/2
       lib/mix/test/test_helper.exs:114: MixTest.Case.in_fixture/3
       lib/mix/test/mix/tasks/compile_test.exs:44: (test)

Reply:
Maybe that:

Enum.reduce({:noop, []}, &merge_diagnostics/2)

Reply:
Oh, I see. Yes, `|> Enum.reduce({:noop, []}, &merge_diagnostics/2)` should work!